Collatz conjecture

TheRealOwenRees opened the pull request:
- Ported the exercise Collatz Conjecture to OCaml.
- Added the `add-practice-exercise` script to generate new practice exercise

Issues:
- environment does not build when following the readme file, with package conflicts
- https://exercism.org/docs/tracks/ocaml/installation works, but then some exercises do not run tests due to missing packages that require base compiler v5.1.0 
- in hindsight, using `configlet` would have been better to create the exercise

BethanyG commented:
@kahgoh  - I triggered the CI and gave a once-over to the files, which look fine to my eye.  

But the ocaml code needs a more experienced once-over than I can really give it.  

Looks like for PR #539, you were talking about making changes to the test generator.  Would that also be something you would want here?  If so, I think @TheRealOwenRees is more than willing, but would need some guidance on the details.  😄

TheRealOwenRees commented:
> > Thanks for raising this PR to add collatz conjecture for the OCaml track! I've got some comments below 👇🏻
> 
> Hello and thank you for your comments. I will look into those and resolve those issues.
> 
> However I cannot run the commands in the Readme as the OCaml environment does not build. I have a local switch for this and running compiler 5.1.0, as well as trying 5.1.1. I get the following errors:
> 
> ```
> [ERROR] Package conflict!
>   * No agreement on the version of ocaml:
>     - (invariant) → ocaml-base-compiler = 5.1.1 → ocaml = 5.1.1
>     - getopts → lemonade >= 0.2.1 → mixture < 0.2.1 → ocaml < 5.0.0
>     You can temporarily relax the switch invariant with `--update-invariant'
>   * No agreement on the version of ocaml-base-compiler:
>     - (invariant) → ocaml-base-compiler = 5.1.1
>     - getopts → lemonade >= 0.2.1 → mixture < 0.2.1 → ocaml < 5.0.0 → ocaml-base-compiler (= 4.08.1 | = 4.11.1)
>   * Incompatible packages:
>     - (invariant) → ocaml-base-compiler = 5.1.1
>     - getopts → lemonade >= 0.2.1 → mixture < 0.2.1 → ocaml < 5.0.0 → ocaml-system >= 4.02.4
>   * Incompatible packages:
>     - (invariant) → ocaml-base-compiler = 5.1.1
>     - getopts → lemonade >= 0.2.1 → mixture < 0.2.1 → ocaml < 5.0.0 → ocaml-variants (< 4.05.2~ | >= 4.08.1)
>   * Missing dependency:
>     - getopts → lemonade >= 0.2.1 → mixture < 0.2.1 → ocaml < 5.0.0 → ocaml-variants >= 4.08.1 → ocaml-beta
>     unmet availability conditions: 'enable-ocaml-beta-repository'
>   * Missing dependency:
>     - getopts → lemonade >= 0.2.1 → mixture < 0.2.1 → ocaml < 5.0.0 → ocaml-variants >= 4.08.1 → system-msvc
>     unmet availability conditions: 'os = "win32"'
> 
> No solution found, exiting
> ```
> 
> I have also tried with the `--update-invariant` option.
> 
> If you have any pointers to get this working, I am certain I can finalise this PR.
> 
> Thanks

I believe to have fixed this. I will submit a PR for troubleshooting / installation guide updates another time, after finalising this PR.

TheRealOwenRees commented:
- removed mistakenly added files
- added template to test generator

The test generator command does not generate the required test for me. I have added the problem specification submodule. 

Could this be that I need a special case for the test, where `result` is a return type with `Ok` and `Error`? I think that might be a little beyond me at this stage.

TheRealOwenRees commented:
Test template added.

However it is generating multiple tests for `0` and `-15` according to what is in the canonical data:
```
 {
      "uuid": "7d4750e6-def9-4b86-aec7-9f7eb44f95a3",
      "description": "zero is an error",
      "property": "steps",
      "input": {
        "number": 0
      },
      "expected": {
        "error": "Only positive numbers are allowed"
      }
    },
    {
      "uuid": "2187673d-77d6-4543-975e-66df6c50e2da",
      "reimplements": "7d4750e6-def9-4b86-aec7-9f7eb44f95a3",
      "description": "zero is an error",
      "comments": ["Collatz Conjecture holds only for positive integers"],
      "property": "steps",
      "input": {
        "number": 0
      },
      "expected": {
        "error": "Only positive integers are allowed"
      }
    },
    ```
    
 If someone can point me in the direction to fix that, I believe I may have finally cracked it.
 
 Thanks

kahgoh commented:
The test generator doesn't handle the `reimplements` 😅. Admittedly, this is also causing some of the broken test generation issues. 

If you're up fixing this limitation that would be greatly appreciated 😁. You'll probably need to dive into the test-generator code. I think the points of interest are:
- `bin_test_gen/test_gen.ml` - the generator's starting point
- `lib_generator/controller.ml` - the controller used by test_gen to apply the templates
- `lib_generator/exercism.ml` - the `of_candidate` function is where the canonical data for the exercise is read

If not, that's okay too 😉 . In this case, let's just mark the test generation broken by adding a `templates/collatz-conjecture/.broken` file. Contents can be empty or a note about not handling the reimplements key.

kahgoh commented:
Also, Github is reporting a merge conflict on `config.json`. Would you mind updating and resolving?

kahgoh commented on node_modules/.lsp/debug.log:
This file has been accidentally added again 😅.

kahgoh replied:
Actually, I was kinda curious what might be generated and added .... At least in my environment, I haven't seen `node_modules` get generated, although my OCaml version is 4.14.2...

TheRealOwenRees replied:
I am running a 5.1.1 global switch for this as the Readme stipulates OCaml 5.1 is needed. Perhaps this is the cause of the issue!

kahgoh replied:
>  Readme stipulates OCaml 5.1 is needed

Oh that's interesting! I had my environment set up a long time ago based on the [track docs](https://exercism.org/docs/tracks/ocaml/installation) which says 4.08.0 😅. We should fix that - I'll add that as a separate issue later.

kahgoh commented:
>  Readme stipulates OCaml 5.1 is needed

Oh that's interesting! I had my environment set up a long time ago based on the [track docs](https://exercism.org/docs/tracks/ocaml/installation) which says 4.08.0 😅. We should fix that - I'll add that as a separate issue later.

kahgoh reviewed (approved):
Thanks, this looks good!
